### ISSUE TITLE
FEAT: IAM Support for AWS RDS (Postgres)

### ISSUE THREAD

**Issue by ndom91** [beekeeper-studio/beekeeper-studio#1443]
**Is your feature request related to a problem? Please describe.**

I see recently IAM Authentication support has been merged for AWS Redshift (https://github.com/beekeeper-studio/beekeeper-studio/pull/1346). But it doesn't look like its available when selecting `Postgres` as the connection type.

I see someone mentioned in that PR about adding it for RDS already (https://github.com/beekeeper-studio/beekeeper-studio/pull/1346#issuecomment-1217164645), but nothing more seems to have been talked about or done in that regard.

Is that something that's on the roadmap? How much work would you guesstimate it would require? I can take a stab at it potentially.

**Describe the solution you'd like**

Ability to login to AWS RDS for Postgres with IAM credentials.

**Database type, if appropriate**

- Postgres

**Comment by rathboma** [comment]
Hey! Honestly I have no idea. I don’t have a database I access that way yet. I’ll take a look to scope it out.

**Comment by ndom91** [comment]
I have a PR, or set of PRs probably, going to add this. The first one should be up soon.

1. [ndom91/add-general-iamoptions](https://github.com/ndom91/beekeeper-studio/tree/ndom91/add-general-iamoptions) - To modify the `redshiftIamOptions` into general iam options and apply them (potentially) to all db types. 
2. WIP

**Comment by MasterOdin** [comment]
To support RDS Postgres, you just need to get a valid `credentials` object which has an `accessKeyId` and `secretAccessKey` (optionally also `sessionToken`), or can use the [`@aws-sdk/credential-provider-ini`](https://www.npmjs.com/package/@aws-sdk/credential-provider-ini) package (or the more publicaly documented [`@aws-sdk/credential-provider-node`](https://www.npmjs.com/package/@aws-sdk/credential-provider-node) package) to parse the credentials from the AWS profile file (given the profile name, or use the default).

Given the credentials, you can then use the [`@aws-sdk/rds-signer`](https://www.npmjs.com/package/@aws-sdk/rds-signer) package like so:

```
const signer = new Signer({
  credentials,
  hostname,
  port,
  region,
  username,
});

const password = await signer.getAuthToken();
```

Could also just go through the `aws-cli` as well, but that would be much more annoying, though it is necessary if you want to support AWS SSO like [dbeaver does.](https://dbeaver.com/docs/wiki/AWS-SSO/).

**Comment by ndom91** [comment]
Ah yeah I was struggling with the CLI route mentioned above. Thanks for the tips, will give that a shot :+1:

**Comment by ndom91** [comment]
Made a little bit of progress, got the aws `~/.aws/credentials` values in (using the `fromNodeProviderChain` which supports all sorts of credential options like env vars as well) and am passing it to the `rds-signer`, but it keeps throwing some odd errors about Object not being a function (duh :sweat_smile:), seems like its trying to call the `credentials` object I'm passing the signer as a function.. 

[`apps/studio/src/lib/db/clients/postgresql.ts`](https://github.com/beekeeper-studio/beekeeper-studio/compare/master...ndom91:beekeeper-studio:ndom91/add-general-iamoptions#diff-4ccaa2470218e19c4b5984257c2e176426baf2cc57e58636269b46e0137b2b59)

Anyway, will do some more digging in the next days/weeks as I have time.

**Comment by MasterOdin** [comment]
The `fromNodeProviderChain` function returns a variable with the type signature `() => Promise<Credentials>` (which is also `CredentialsProvider` in `@types/aws`) so the expectation is that `Signer` _should_ be calling it.

**Comment by ndom91** [comment]
Yeah so I saw that it takes the object, a promise that resolves to that object or a function returning either. I'd have to double check, but I'm pretty sure I was simply passing the promise which resolved to the `{ clientId, clientSecret }` object. I'll ahve to double check though, thanks for the pointer.

**Comment by rathboma** [comment]
Thanks @MasterOdin for helping out here! I'm totally new to this also

**Comment by MasterOdin** [comment]
Since the `Signer` class doesn't actually make any HTTP calls, it should be possible to easily validate your approach works via the REPL:

```
> const { fromNodeProviderChain } = require('@aws-sdk/credential-providers')
undefined
> const { Signer } = require('@aws-sdk/rds-signer');
undefined
> signer = new Signer({
... credentials: fromNodeProviderChain(),
... region: 'eu-west-1',
... hostname: 'localhost',
... port: 5849,
... username: 'foo',
... });
Signer {
  protocol: 'https:',
  service: 'rds-db',
  credentials: [AsyncFunction (anonymous)],
  hostname: 'localhost',
  port: 5849,
  region: 'eu-west-1',
  sha256: [Function: bound Hash],
  username: 'foo'
}
> await signer.getAuthToken()
'<redacted>'
```

So `fromNodeProviderChain()` does work fine here for the `credentials`. My only thought would be maybe one of the other parameters that is getting passed in is incorrect in some way? Alternatively, what is the `signer` object that gets printed out in your code and how does it compare the above?

You could modify the use of signer to do this as well to avoid the promise resolution inside the signer altogether:

```
signer = new Signer({
  credentials: await fromNodeProviderChain()(),
  // ...
});
```

and that'd at least rule out that that function is the root cause? Of course, it may be that `fromNodeProviderChain` cannot actually find any credentials on your computer and it's throwing an error because of it...

**Comment by ndom91** [comment]
Yeah so turns out it doesn't return a promise

![image](https://user-images.githubusercontent.com/7415984/219509353-0d68bd5f-1ba4-4ea5-b9f2-d3be743da990.png)

So I just set that to a variable and tried to log it and then pass it to the `credentials` parameter and it just dies on the `fromNodeProviderChain()` call with this error:

![image](https://user-images.githubusercontent.com/7415984/219508884-f00b2494-f1b2-4163-8322-e2b3e66dad56.png)

That last `CREDENTIALPROVIDER` logged line is where this snippet of code begins:

```js
console.log('** CREDENTIALPROVIDER')

const nodeProviderChainCredentials = fromNodeProviderChain()

console.log('** CREDENTIALPROVIDER NODE CHAIN', nodeProviderChainCredentials)

const signer = new Signer({
  credentials: nodeProviderChainCredentials,
  region: 'eu-west-1',
  hostname: server.config.host,
  port: server.config.port,
  username: server.config.user
});

console.log('** SIGNER', signer)
const token = await signer.getAuthToken()
console.log('** TOKEN', token)
passwordResolver = async () => {
  return token
}
```

It never gets to logging `signer` or `token`.

I definitely do have an `~/.aws/credentials` file with a `default` profile setup whose credentials should allow me access to the testing RDS Postgres database I have setup here.

:thinking:

**Comment by ndom91** [comment]
And that `AwsCredentialIdentityProvider` is apparenlty just the object we thoguht it was :thinking: 

![image](https://user-images.githubusercontent.com/7415984/219509659-31304695-9e06-49b9-89ca-d819c5a59fd5.png)

**Comment by ndom91** [comment]
Okay well I figured out part of the problem.. apparently the import is `undefined`

Docs say this should work, but isn't for me :joy:

```
import { fromNodeProviderChain } from "@aws-sdk/credential-providers";

console.log(fromNodeProviderchain) // undefined
```

**Comment by ndom91** [comment]
Alright so digging into the `@aws-sdk/credential-providers` in my `node_modules`, showed that it _was_ there and there was a named export called `fromNodeProviderChain`, so no idea what was going on there.. I tried wiping the node_modules and re-installing to no avail :/

However, I was able to `import { fromIni } from `@aws-sdk/credential-provider-ini` for now. This finally got the credentials picked up and the signer/token to be generated.

Unfortunately I wasn't able to successfully authenticate yet though. `pg_hbf.conf rejects host...`. I don't have another DB client that supports IAM auth to test this on, but similar credentials on this DB work for all of my coworkers, so im 99% sure this should work. Will double check everything :crossed_fingers:

**Comment by ndom91** [comment]
Final update: just needed SSL enabled and it worked :joy:

Will clean up the PR and push final soon 

https://github.com/beekeeper-studio/beekeeper-studio/compare/master...ndom91:beekeeper-studio:ndom91/add-general-iamoptions

**Comment by ndom91** [comment]
PR is up :partying_face: 

https://github.com/beekeeper-studio/beekeeper-studio/pull/1514

**Comment by mtesch-um** [comment]
maybe this can be closed because PR2330
